control/control.go

```diff
@@ -366,12 +366,12 @@ func (c *Controller) Solve(ctx context.Context, req *controlapi.SolveRequest) (*
 	}
 
 	var expis []exporter.ExporterInstance
-	for _, ex := range req.Exporters {
+	for i, ex := range req.Exporters {
 		exp, err := w.Exporter(ex.Type, c.opt.SessionManager)
 		if err != nil {
 			return nil, err
 		}
-		expi, err := exp.Resolve(ctx, ex.ID, ex.Attrs)
+		expi, err := exp.Resolve(ctx, i, ex.Attrs)
```

Comment:
Hmm, this assumes that all exporters will use sync targets, but at least for the image exporter it's not the case. I think the easiest test is to re-arrange the exports in the test:

```go
Exports: []ExportEntry{
	{
		Type: imageExporter,
		Attrs: map[string]string{
			"name": target1,
		},
	},
	{
		Type: imageExporter,
		Attrs: map[string]string{
			"name":           target2,
			"oci-mediatypes": "true",
		},
	},
	// Ensure that multiple local exporter destinations are written properly
	{
		Type:      ExporterLocal,
		OutputDir: destDir,
	},
	{
		Type:      ExporterLocal,
		OutputDir: destDir2,
	},
	// Ensure that multiple instances of the same exporter are possible
	{
		Type:   ExporterTar,
		Output: fixedWriteCloser(outW),
	},
	{
		Type:   ExporterTar,
		Output: fixedWriteCloser(outW2),
	},
},
```
e.g. move the imge exporter to the beginning.

Reply:
If we have the image exporters in front, there will be a gap for sync targets (they will not start with `0`), so one cannot really assume the order in [NewFSSyncTarget](https://github.com/moby/buildkit/compare/master...jedevc:buildkit:enable-multi-exporters#diff-7078d905e2386549f621b6a42a2be0947615d2b0d51b1c0acb5f13caaf126ac6R264):
```go
for id, t := range targets {
```
`id` will need to start with `len(imageExporters)` in this contrived example. Of course, it's just one of the combinations.

Reply:
I'm still confused - the IDs in this case will start from `2`, but this isn't really an issue IMO - all that matters is that the IDs are unique.

Reply:
Do we agree that by shuffling the order of exporters, the multiple exporter testcase should not fail? E.g. both of these configurations should not result in test failures:
```go
Exports: []ExportEntry{
	{
		Type: imageExporter,
		// Other attrs omitted
	},
	{Type: ExporterLocal},
	{Type: ExporterLocal},
	{Type: imageExporter},
	{Type: ExporterTar},
	{Type: ExporterTar},
}
```
and this:
```go
Exports: []ExportEntry{
	{
		Type: imageExporter,
		// Other attrs omitted
	},
	{Type: imageExporter},
	{Type: ExporterLocal},
	{Type: ExporterLocal},
	{Type: ExporterTar},
	{Type: ExporterTar},
}
```
as they do now.

What I'm referring to is the way the [syncTargets](https://github.com/jedevc/buildkit/blob/enable-multi-exporters/client/solve.go#L151) slice is being built.
Instead, the exporters could be shuffled to move the local exporters to the front so their order is deterministic, e.g. by doing something like this:
```go
// Push the local exporters to the front to build
// stable ID order for file sync server
var localExports, imageExports []ExportEntry
for _, exp := range opt.Exports {
	if exp.Type != ExporterImage {
		localExports = append(localExports, exp)
	} else {
		imageExports = append(imageExports, exp)
	}
}
opt.Exports = append(localExports, imageExports...)

var syncTargets []filesync.FSSyncTarget
for _, ex := range opt.Exports {
```